File: 백준/Gold/1562. 계단 수/계단 수.c
#include <stdio.h>
#include <stdlib.h>

#define MOD		1000000000
#define DIGIT	10
#define EMPTY	-1

int _n;
long long **_stair;
long long ****_dp;

int
get_min(int val1, int val2)
{
	return val1 < val2 ? val1 : val2;
}

int
get_max(int val1, int val2)
{
	return val1 > val2 ? val1 : val2;
}

long long
get_stair(int idx, int num)
{
	long long sum;

	if (_stair[idx][num] != EMPTY)
		return _stair[idx][num];
	if (idx == _n - 1)
		return _stair[idx][num] = 1;
	sum = 0;
	if (num > 0)
		sum = (sum + get_stair(idx + 1, num - 1)) % MOD;
	if (num < 9)
		sum = (sum + get_stair(idx + 1, num + 1)) % MOD;
	return _stair[idx][num] = sum;
}

long long
get_dp(int idx, int max, int min, int num)
{
	long long sum;

	if (_dp[idx][max][min][num] != EMPTY)
		return _dp[idx][max][min][num];
	if (max == 9 && min == 0)
		return _dp[idx][max][min][num] = 0;
	if (idx == _n - 1)
		return _dp[idx][max][min][num] = 1;
	sum = 0;
	if (num > 0)
		sum = (sum + get_dp(idx + 1, max, get_min(min, num - 1), num - 1)) % MOD;
	if (num < 9)
		sum = (sum + get_dp(idx + 1, get_max(max, num + 1), min, num + 1)) % MOD;
	return _dp[idx][max][min][num] = sum;
}

int
main()
{
	int i;
	int j;
	int k;
	int l;
	long long sum;

	scanf("%d", &_n);
	_dp = (long long ****) malloc(_n * sizeof(long long ***));
	for (i = 0; i < _n; i++) {
		_dp[i] = (long long ***) malloc(DIGIT * sizeof(long long **));
		for (j = 0; j < DIGIT; j++) {
			_dp[i][j] = (long long **) malloc((j + 1) * sizeof(long long *));
			for (k = 0; k <= j; k++) {
				_dp[i][j][k] = (long long *) malloc(DIGIT * sizeof(long long));
				for (l = 0; l < DIGIT; l++)
					_dp[i][j][k][l] = EMPTY;
			}
		}
	}
	_stair = (long long **) malloc(_n * sizeof(long long *));
	for (i = 0; i < _n; i++) {
		_stair[i] = (long long *) malloc(DIGIT * sizeof(long long));
		for (j = 0; j < DIGIT; j++)
			_stair[i][j] = EMPTY;
	}
	sum = 0;
	for (i = 1; i < DIGIT; i++)
		sum = (sum + get_stair(0, i)) % MOD;
	for (i = 1; i < DIGIT; i++)
		sum = (sum - get_dp(0, i, i, i)) % MOD;
	if (sum < 0)
		sum += MOD;
	printf("%d", (int) sum);
	return 0;
}

File: 백준/Gold/1562. 계단 수/계단 수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    private static final int MOD = 1_000_000_000;
    private static final int DIGIT = 10;
    private static final int HALF = DIGIT >> 1;
    private static final int EMPTY = -1;
    private static final int NONE = 0;
    private static final int ZERO = 1;
    private static final int NINE = 1 << 1;
    private static final int BOTH = ZERO | NINE;

    private static long[][] stair;
    private static long[][][] dp;

    private static long getStair(int idx, int num) {
        long sum;

        if (num >= HALF) {
            num = DIGIT - num - 1;
        }
        if (stair[idx][num] != EMPTY) {
            return stair[idx][num];
        }
        sum = 0;
        if (num > 0) {
            sum = (sum + getStair(idx + 1, num - 1)) % MOD;
        }
        sum = (sum + getStair(idx + 1, num + 1)) % MOD;
        return stair[idx][num] = sum;
    }

    private static long getDp(int idx, int num, int bit) {
        long sum;

        if (num >= HALF) {
            num = DIGIT - num - 1;
            bit = ((bit & ZERO) << 1) | ((bit & NINE) >> 1);
        }
        if (dp[idx][num][bit] != EMPTY) {
            return dp[idx][num][bit];
        }
        sum = 0;
        if (num == 1) {
            sum = (sum + getDp(idx + 1, 0, bit | ZERO)) % MOD;
        } else if (num > 1) {
            sum = (sum + getDp(idx + 1, num - 1, bit)) % MOD;
        }
        sum = (sum + getDp(idx + 1, num + 1, bit)) % MOD;
        return dp[idx][num][bit] = sum;
    }

    public static void main(String[] args) throws IOException {
        int n;
        int i;
        int j;
        int k;
        int end;
        long sum;
        long ans;
        BufferedReader br;

        br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        stair = new long[n][HALF];
        dp = new long[n][HALF][BOTH + 1];
        end = n - 1;
        for (i = 0; i < end; i++) {
            for (j = 0; j < HALF; j++) {
                stair[i][j] = EMPTY;
                for (k = 0; k < BOTH; k++) {
                    dp[i][j][k] = EMPTY;
                }
            }
        }
        dp[end][0][ZERO] = 1;
        stair[end][0] = 1;
        for (i = 1; i < HALF; i++) {
            stair[end][i] = 1;
            for (j = 0; j < BOTH; j++) {
                dp[end][i][j] = 1;
            }
        }
        sum = 0;
        for (i = 0; i < HALF; i++) {
            sum = (sum + getStair(0, i)) % MOD;
        }
        sum = (sum << 1) % MOD;
        ans = (sum - getStair(0, 0)) % MOD;
        sum = getDp(0, 0, ZERO) % MOD;
        for (i = 1; i < HALF; i++) {
            sum = (sum + getDp(0, i, NONE)) % MOD;
        }
        sum = (sum << 1) % MOD;
        ans = (ans - ((sum - getDp(0, 0, ZERO)) % MOD)) % MOD;
        System.out.print(ans < 0 ? ans + MOD : ans);
    }
}
